File: src/org/kde/kdeconnect/async/BackgroundJobHandler.java
/*
 * SPDX-FileCopyrightText: 2018 Erik Duisters <e.duisters1@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
 */

package org.kde.kdeconnect.async;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import androidx.annotation.Nullable;

/**
 * Scheduler for {@link BackgroundJob} objects.
 * <p>
 *     We use an internal {@link ThreadPoolExecutor} to catch Exceptions and
 *     pass them along to {@link #handleUncaughtException(Future, Throwable)}.
 * </p>
 */
public class BackgroundJobHandler {
    private static final String TAG = BackgroundJobHandler.class.getSimpleName();

    private final Map<BackgroundJob, Future<?>> jobMap = new HashMap<>();
    private final Object jobMapLock = new Object();

    private class MyThreadPoolExecutor extends ThreadPoolExecutor {
        MyThreadPoolExecutor(int corePoolSize, int maxPoolSize, long keepAliveTime, TimeUnit unit, BlockingQueue<Runnable> workQueue) {
            super(corePoolSize, maxPoolSize, keepAliveTime, unit, workQueue);
        }

        @Override
        protected void afterExecute(Runnable r, Throwable t) {
            super.afterExecute(r, t);

            if (!(r instanceof Future)) {
                return;
            }

            Future<?> future = (Future<?>) r;

            if (t == null) {
                try {
                    future.get();
                } catch (CancellationException ce) {
                    Log.d(TAG,"afterExecute got a CancellationException");
                } catch (ExecutionException ee) {
                    t = ee;
                } catch (InterruptedException ie) {
                    Log.d(TAG, "afterExecute got an InterruptedException");
                    Thread.currentThread().interrupt();    // ignore/reset
                }
            }

            if (t != null) {
                BackgroundJobHandler.this.handleUncaughtException(future, t);
            }
        }
    }

    private final ThreadPoolExecutor threadPoolExecutor;
    private Handler handler;

    private BackgroundJobHandler(int corePoolSize, int maxPoolSize, long keepAliveTime, TimeUnit unit, BlockingQueue<Runnable> workQueue) {
        this.handler = new Handler(Looper.getMainLooper());
        this.threadPoolExecutor = new MyThreadPoolExecutor(corePoolSize, maxPoolSize, keepAliveTime, unit, workQueue);
    }

    public void runJob(BackgroundJob bgJob) {
        Future<?> f;

        bgJob.setBackgroundJobHandler(this);

        try {
            synchronized (jobMapLock) {
                f = threadPoolExecutor.submit(bgJob);
                jobMap.put(bgJob, f);
            }
        } catch (RejectedExecutionException e) {
            Log.d(TAG,"threadPoolExecutor.submit rejected a background job: " + e.getMessage());

            bgJob.reportError(e);
        }
    }

    public boolean isRunning(long jobId) {
        synchronized (jobMapLock) {
            for (BackgroundJob job : jobMap.keySet()) {
                if (job.getId() == jobId) {
                    return true;
                }
            }
        }

        return false;
    }

    @Nullable
    public BackgroundJob getJob(long jobId) {
        synchronized (jobMapLock) {
            for (BackgroundJob job : jobMap.keySet()) {
                if (job.getId() == jobId) {
                    return job;
                }
            }
        }

        return null;
    }

    void cancelJob(BackgroundJob job) {
        synchronized (jobMapLock) {
            if (jobMap.containsKey(job)) {
                Future<?> f = jobMap.get(job);

                if (f.cancel(true)) {
                    threadPoolExecutor.purge();
                }

                jobMap.remove(job);
            }
        }
    }

    private void handleUncaughtException(Future<?> ft, Throwable t) {
        synchronized (jobMapLock) {
            for (Map.Entry<BackgroundJob, Future<?>> pairs : jobMap.entrySet()) {
                Future<?> future = pairs.getValue();

                if (future == ft) {
                    pairs.getKey().reportError(t);
                    break;
                }
            }
        }
    }

    void onFinished(BackgroundJob job) {
        synchronized (jobMapLock) {
            jobMap.remove(job);
        }
    }

    void runOnUiThread(Runnable runnable) {
        handler.post(runnable);
    }

    public static BackgroundJobHandler newFixedThreadPoolBackgroundJobHander(int numThreads) {
        return new BackgroundJobHandler(numThreads, numThreads, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>());
    }
}

File: src/org/kde/kdeconnect/async/BackgroundJobHandler.kt
/*
 * SPDX-FileCopyrightText: 2018 Erik Duisters <e.duisters1@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only OR GPL-3.0-only OR LicenseRef-KDE-Accepted-GPL
 */
package org.kde.kdeconnect.async

import android.os.Handler
import android.os.Looper
import android.util.Log
import java.util.concurrent.BlockingQueue
import java.util.concurrent.CancellationException
import java.util.concurrent.ExecutionException
import java.util.concurrent.Future
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.RejectedExecutionException
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

/**
 * Scheduler for [BackgroundJob] objects.
 *
 * We use an internal [ThreadPoolExecutor] to catch Exceptions and pass them along to [.handleUncaughtException].
 *
 * Might be able to be replaced with coroutines later
 */
class BackgroundJobHandler {
    private constructor(corePoolSize: Int, maxPoolSize: Int, keepAliveTime: Long, unit: TimeUnit, workQueue: BlockingQueue<Runnable>) {
        this.threadPoolExecutor = MyThreadPoolExecutor(corePoolSize, maxPoolSize, keepAliveTime, unit, workQueue)
        this.handler = Handler(Looper.getMainLooper())
    }

    private val jobMap: MutableMap<BackgroundJob<*, *>, Future<*>> = HashMap()
    private val jobMapLock: Any = Any()

    private inner class MyThreadPoolExecutor(corePoolSize: Int, maxPoolSize: Int, keepAliveTime: Long, unit: TimeUnit, workQueue: BlockingQueue<Runnable>) : ThreadPoolExecutor(corePoolSize, maxPoolSize, keepAliveTime, unit, workQueue) {
        override fun afterExecute(runnable: Runnable, throwable: Throwable?) {
            super.afterExecute(runnable, throwable)

            if (runnable !is Future<*>) {
                return
            }

            val future = runnable as Future<*>

            var t: Throwable? = throwable
            if (t == null) {
                try {
                    future.get()
                }
                catch (ce: CancellationException) {
                    Log.d(LOG_TAG, "afterExecute got a CancellationException")
                }
                catch (ee: ExecutionException) {
                    t = ee
                }
                catch (ie: InterruptedException) {
                    Log.d(LOG_TAG, "afterExecute got an InterruptedException")
                    Thread.currentThread().interrupt() // ignore / reset
                }
            }

            if (t != null) {
                this@BackgroundJobHandler.handleUncaughtException(future, t)
            }
        }
    }

    private val threadPoolExecutor: ThreadPoolExecutor
    private val handler: Handler

    fun runJob(bgJob: BackgroundJob<*, *>) {
        bgJob.setBackgroundJobHandler(this)
        try {
            synchronized(jobMapLock) {
                val future: Future<*> = threadPoolExecutor.submit(bgJob)
                jobMap.put(bgJob, future)
            }
        }
        catch (e: RejectedExecutionException) {
            Log.d(LOG_TAG, "threadPoolExecutor.submit rejected a background job: ${e.message}")
            bgJob.reportError(e)
        }
    }

    fun isRunning(jobId: Long): Boolean {
        synchronized(jobMapLock) {
            return jobMap.keys.any { it.id == jobId }
        }
    }

    fun getJob(jobId: Long): BackgroundJob<*, *>? {
        synchronized(jobMapLock) {
            return jobMap.keys.find { it.id == jobId }
        }
    }

    fun cancelJob(job: BackgroundJob<*, *>) {
        synchronized(jobMapLock) {
            if (jobMap.containsKey(job)) {
                if (jobMap[job]!!.cancel(true)) {
                    threadPoolExecutor.purge()
                }
                jobMap.remove(job)
            }
        }
    }

    private fun handleUncaughtException(ft: Future<*>, t: Throwable) {
        synchronized(jobMapLock) {
            jobMap.entries.find { it.value == ft }?.key?.reportError(t)
        }
    }

    fun onFinished(job: BackgroundJob<*, *>) {
        synchronized(jobMapLock) {
            jobMap.remove(job)
        }
    }

    fun runOnUiThread(runnable: Runnable) {
        handler.post(runnable)
    }

    companion object {
        private val LOG_TAG: String = BackgroundJobHandler::class.java.simpleName

        @JvmStatic
        fun newFixedThreadPoolBackgroundJobHandler(numThreads: Int): BackgroundJobHandler {
            return BackgroundJobHandler(numThreads, numThreads, 0L, TimeUnit.MILLISECONDS, LinkedBlockingQueue())
        }
    }
}
